src/parser2.rs

```diff
@@ -0,0 +1,86 @@
+
+
+impl_rdp! {
+    grammar! {
+        // basic blocks of the language
+        op_or         = { ["or"] }
+        op_and        = { ["and"] }
+        op_comparison = { ["<="] | [">="] | ["<"] | [">"] | ["!="] | ["=="] }
+        op_add_sub    = { ["+"] | ["-"] }
+        op_mul_div    = { ["*"] | ["/"] }
+        boolean       = { ["true"] | ["false"] }
+
+        decimals = @{ ["."] ~ ['0'..'9']+ }
+        int      = @{ ["-"]? ~ (["0"] | ['1'..'9'] ~ ['0'..'9']*) }
+        float    = @{ ["-"]? ~ (["0"] ~ decimals? | ['1'..'9'] ~ ['0'..'9']* ~ decimals?) }
+
+        identifier = @{
+            (['a'..'z'] | ['A'..'Z'] | ["_"]) ~
+            (['a'..'z'] | ['A'..'Z'] | ["_"] | ["."] | ['0'..'9'])*
+        }
+        // matches anything between 2 double quotes
+        string = @{ ["\""] ~ (!(["\""]) ~ any )* ~ ["\""]}
+
+        // Precedence climbing
+        expression = _{
+            { expression | identifier | int | float }
+            or          = { op_or }
+            and         = { op_and }
+            comparison  = { op_comparison }
+            add_sub     = { op_add_sub }
+            mul_div_mod = { op_mul_div }
+        }
+
+        // Tera specific things
+
+        // different types of blocks
+        variable_start = { ["{{"] }
+        variable_end   = { ["}}"] }
+        tag_start      = { ["{%"] }
+        tag_end        = { ["%}"] }
+        comment_start  = { ["{#"] }
+        comment_end    = { ["#}"] }
+        block_start    = { variable_start | tag_start | comment_start }
+
+        // Actual tags
+        extends_tag  = { tag_start ~ ["extends"] ~ string ~ tag_end }
+
+        comment_tag  = { comment_start ~ (!comment_end ~ any )* ~ comment_end }
+        block_tag    = { tag_start ~ ["block"] ~ identifier ~ tag_end }
+        endblock_tag = { tag_start ~ ["endblock"] ~ identifier ~ tag_end }
+        if_tag       = { tag_start ~ ["if"] ~ expression ~ tag_end }
+        elif_tag     = { tag_start ~ ["elif"] ~ expression ~ tag_end }
+        else_tag     = { tag_start ~ ["else"] ~ tag_end }
+        endif_tag    = { tag_start ~ ["endif"] ~ tag_end }
+        for_tag      = { tag_start ~ ["for"] ~ identifier ~ ["in"] ~ identifier ~ tag_end }
+        endfor_tag   = { tag_start ~ ["endfor"] ~ tag_end }
+
+        text = { (!(block_start) ~ any )* }
+
+        content = {
+            text |
+            comment_tag |
+            block_tag ~ content ~ endblock_tag |
+            if_tag ~ content ~ (elif_tag ~ content)* ~ (else_tag ~ content)? ~ endif_tag |
+            for_tag ~ content ~ endfor_tag
+        }
+
+        // top level node
+        template = { extends_tag? ~ content ~ eoi }
+
+        whitespace = _{ [" "] | ["\t"] | ["\u{000C}"] | ["\r"] | ["\n"] }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use pest::prelude::*;
+    use super::Rdp;
+
+    #[test]
+    fn test_only_text() {
+        let mut parser = Rdp::new(StringInput::new("Hello\n 世界"));
+        assert!(parser.expression());
```

Comment:
I'd recommend to update your examples/doc to use

```rust
assert_eq!(parser.queue(), &queue);
```

rather than
```rust
assert!(parser.queue().iter().eq(&queue));
```